src/packet/bwe/time.rs

```diff
@@ -0,0 +1,739 @@
+use std::ops::{Add, AddAssign, Div, Neg as _, Sub, SubAssign};
+use std::time::{Duration, Instant};
+
+use crate::bwe::Bitrate;
+use crate::rtp_::DataSize;
+
+/// Wrapper for [`Instant`] that provides additional time points in the past or future.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(super) enum SuperInstant {
+    /// A time in the past that already happened.
+    DistantPast,
+
+    /// An exact instant.
+    Exact(Instant),
+
+    /// A time in the future that will never happen.
+    DistantFuture,
+}
+
+/// Wrapper for [`Duration`] that can be negative and provides a duration to a
+/// distant future or past.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(super) enum TimeDelta {
+    /// Time delta to some event in distant past that already happened.
+    MinusInf,
+
+    /// An exact negative duration.
+    Negative(Duration),
+
+    /// An exact positive duration.
+    Positive(Duration),
+
+    /// Time delta to some event in distant future that will never happen.
+    PlusInf,
+}
+
+impl TimeDelta {
+    pub(super) const ZERO: Self = Self::Positive(Duration::ZERO);
+
+    /// Returns the number of seconds contained by this [`TimeDelta`] as `f64`.
+    pub fn as_secs_f64(&self) -> f64 {
+        match self {
+            Self::Negative(d) => d.as_secs_f64().neg(),
+            Self::Positive(d) => d.as_secs_f64(),
+            Self::PlusInf => f64::INFINITY,
+            Self::MinusInf => f64::NEG_INFINITY,
+        }
+    }
+}
+
+#[cfg(test)]
+impl TimeDelta {
+    /// Creates a [`TimeDelta`] from seconds.
+    pub const fn from_secs(secs: i64) -> TimeDelta {
+        if secs >= 0 {
+            Self::Positive(Duration::from_secs(secs as u64))
+        } else {
+            Self::Negative(Duration::from_secs(-secs as u64))
+        }
+    }
+
+    /// Creates a [`TimeDelta`] from milliseconds.
+    pub const fn from_millis(millis: i64) -> Self {
+        if millis >= 0 {
+            Self::Positive(Duration::from_millis(millis as u64))
+        } else {
+            Self::Negative(Duration::from_millis(-millis as u64))
+        }
+    }
+}
+
+impl SuperInstant {
+    /// Indicates whether this [`SuperInstant`] is [`SuperInstant::Exact`].
+    pub const fn is_finite(&self) -> bool {
+        matches!(self, Self::Exact(_))
+    }
+
+    /// Indicates whether this [`SuperInstant`] is [`SuperInstant::DistantPast`]
+    /// or [`SuperInstant::DistantFuture`].
+    pub const fn is_not_finite(&self) -> bool {
+        !self.is_finite()
+    }
+}
+
+impl Add<TimeDelta> for SuperInstant {
+    type Output = Self;
+
+    fn add(self, rhs: TimeDelta) -> Self::Output {
+        match (self, rhs) {
+            (Self::DistantFuture, _) | (_, TimeDelta::PlusInf) => Self::DistantFuture,
+            (Self::DistantPast, _) | (_, TimeDelta::MinusInf) => Self::DistantPast,
+            (Self::Exact(i), TimeDelta::Negative(d)) => Self::Exact(i - d),
+            (Self::Exact(i), TimeDelta::Positive(d)) => Self::Exact(i + d),
```

Comment:
Sadly Rust doesn't seem to have `Instant::checked_sub(Duration)`, which would fit neatly into this mechanic. Going via u128 to do it manually seems expensive, so let's leave it like this. Maybe with a comment to look out for future Rust std API.

Reply:
There is an [`Instant::checked_sub(Duration)`](https://doc.rust-lang.org/std/time/struct.Instant.html#method.checked_sub). And what you mean is going `DistantFuture` on overflow and `DistantPast` on underflow instead of panicking? Well, I thought that panicking is a better option since if overflow/underflow happens than this is most likely an unexpected error. But that probably won't ever happen so Im okay with changing it if you want

Reply:
Alright! You thought about it. Let's leave it.